File: v1/src/main/java/com/api/v1/services/CustomerRegistrationServiceImpl.java
package com.api.v1.services;

import com.api.v1.mappers.CreditCardResponseMapper;
import com.api.v1.mappers.CustomerResponseMapper;
import com.api.v1.builders.CustomerBuilder;
import com.api.v1.domain.Customer;
import com.api.v1.domain.CustomerRepository;
import com.api.v1.dtos.CustomerRegistrationRequestDto;
import com.api.v1.dtos.CustomerResponseDto;
import com.api.v1.exceptions.DuplicatedSsnException;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class CustomerRegistrationServiceImpl implements CustomerRegistrationService {

    @Autowired
    private CustomerRepository customerRepository;

    @Override
    public Mono<CustomerResponseDto> register(@Valid CustomerRegistrationRequestDto request) {
        return customerRepository
                .findAll()
                .filter(e -> e.getSsn().equals(request.ssn()))
                .hasElements()
                .flatMap(exists -> {
                    if (exists) return Mono.error(new DuplicatedSsnException(request.ssn()));
                    return Mono.defer(() -> {
                        Customer customer = CustomerBuilder.create().fromDto(request).build();
                        return customerRepository.save(customer);
                    }).flatMap(e -> Mono.just(new CustomerResponseMapper(e).mapToDto()));
                });
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/customer/services/CustomerRegistrationServiceImpl.kt
package com.api.v2.customer.services

import com.api.v2.customer.domain.Customer
import com.api.v2.customer.domain.CustomerRepository
import com.api.v2.customer.dtos.CustomerResponseDto
import com.api.v2.customer.exceptions.DuplicatedSsnException
import com.api.v2.customer.mappers.CustomerResponseMapper
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class CustomerRegistrationServiceImpl: CustomerRegistrationService {

    @Autowired
    lateinit var customerRepository: CustomerRepository

    override suspend fun register(requestDto: @Valid Customer): CustomerResponseDto {
        return withContext(Dispatchers.IO) {
            if (checkTheExistenceOfDuplicatedSsn(requestDto.ssn)) {
                throw DuplicatedSsnException(requestDto.ssn)
            }
            val savedCustomer = customerRepository.save(requestDto)
            CustomerResponseMapper.mapToDto(savedCustomer)
        }
    }

    private suspend fun checkTheExistenceOfDuplicatedSsn(ssn: String): Boolean {
        return customerRepository.findAll().firstOrNull { e -> e.ssn == ssn } != null
    }

}